packages/api/src/tray.ts

```diff
@@ -99,12 +24,30 @@ export interface TrayIconLeaveEvent {
  * - **Linux**: Unsupported. The event is not emitted even though the icon is shown,
  * the icon will still show a context menu on right click.
  */
-export type TrayIconEvent =
-  | { click: TrayIconClickEvent }
-  | { doubleClick: TrayIconDoubleClickEvent }
-  | { enter: TrayIconEnterEvent }
-  | { move: TrayIconMoveEvent }
-  | { leave: TrayIconLeaveEvent }
+export type TrayIconEvent<T extends TrayIconEventType = TrayIconEventType> = {
+  /** The tray icon event type */
+  type: T
+  /** Id of the tray icon which triggered this event. */
+  id: string
+  /** Physical position of the click the triggered this event. */
+  position: PhysicalPosition
+  /** Position and size of the tray icon. */
+  rect: {
+    position: PhysicalPosition
+    size: PhysicalSize
+  }
+} & (T extends 'Click' | 'DoubleClick'
+  ? {
+      /** Mouse button that triggered this event. */
+      button: MouseButton
+    }
+  : never) &
+  (T extends 'Click'
+    ? {
+        /** Mouse button state when this event was triggered. */
+        buttonState: MouseButtonState
+      }
+    : never)
```

Comment:
still not sure how to make these show error in IDE when narrowing type, basically 
```ts
if (event.type === "Leave") {
  e.id; should work
  e.button; // should error
  e.buttonState; shuold error
}
```

Reply:
Try this

```ts
export type BaseTrayIconEvent<T extends TrayIconEventType> = {
  /** The tray icon event type */
  type: T
  /** Id of the tray icon which triggered this event. */
  id: string
  /** Physical position of the click the triggered this event. */
  position: PhysicalPosition
  /** Position and size of the tray icon. */
  rect: {
    position: PhysicalPosition
    size: PhysicalSize
  }
}

export type TrayIconClickEvent = {
  /** Mouse button that triggered this event. */
  button: MouseButton
}

/**
 * Describes a tray icon event.
 *
 * #### Platform-specific:
 *
 * - **Linux**: Unsupported. The event is not emitted even though the icon is shown,
 * the icon will still show a context menu on right click.
 */
export type TrayIconEvent =
  | (BaseTrayIconEvent<'Click'> &
      TrayIconClickEvent & {
        /** Mouse button state when this event was triggered. */
        buttonState: MouseButtonState
      })
  | (BaseTrayIconEvent<'DoubleClick'> & TrayIconClickEvent)
  | BaseTrayIconEvent<'Enter'>
  | BaseTrayIconEvent<'Move'>
  | BaseTrayIconEvent<'Leave'>
```

Reply:
yeah I've tried that and it is still not narrowing correctly

Reply:
oh nvm, I had a typo, thanks for help

Reply:
hmm, it works on my end

![image](https://github.com/user-attachments/assets/7d9e0f87-61bc-43a4-ba72-838da0e5c659)
![image](https://github.com/user-attachments/assets/cb11a3d6-3162-460d-85ed-3ca3b60b855c)

Reply:
yeah it works now, I had `TrayIconEventBase<'Click'>` twice instead of `TrayIconEventBase<'Click'>` and another `TrayIconEventBase<'DoubleClick'`